File: 백준/Silver/2529. 부등호/부등호.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
N = int(input())
signOfInequality = list(input().strip().split())
max_val = 0
min_val = int(1e11)


def dfs(n, array):
    if n == N:
        global max_val, min_val
        max_val = max(max_val, int(''.join([str(i) for i in array])))
        min_val = min(min_val, int(''.join([str(i) for i in array])))
        return
    
    sign = signOfInequality[n]
    lastValue = array[-1]
    
    if sign == "<":
        for currentValue in range(lastValue + 1, 10):
            if lastValue < currentValue and currentValue not in array:
                array.append(currentValue)
                dfs(n+1, array)
                array.pop()
    elif sign == ">":
        for currentValue in range(lastValue):
            if lastValue > currentValue and currentValue not in array:
                array.append(currentValue)
                dfs(n+1, array)
                array.pop()
    return
for i in range(10):
    dfs(0, [i])
    
print(str(max_val).zfill(N+1))
print(str(min_val).zfill(N+1))
            
        

File: 백준/Silver/2529. 부등호/부등호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    static int N;
    static char[] signOfInequality;
    static long maxVal = 0;
    static long minVal = Long.MAX_VALUE;
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        signOfInequality = br.readLine().replaceAll("\\s+","").toCharArray();

        for (int i = 0; i < 10; i++) {
            dfs(0, i, new boolean[10], "" + i);
        }

        System.out.println(String.format("%0" + (N + 1) + "d", maxVal));
        System.out.println(String.format("%0" + (N + 1) + "d", minVal));
    }

    static void dfs(int index, int lastValue, boolean[] visited, String number) {
        if (index == N) {
            long num = Long.parseLong(number);
            maxVal = Math.max(maxVal, num);
            minVal = Math.min(minVal, num);
            return;
        }

        char sign = signOfInequality[index];
        visited[lastValue] = true;

        if (sign == '<') {
            for (int currentValue = lastValue + 1; currentValue < 10; currentValue++) {
                if (!visited[currentValue]) {
                    dfs(index + 1, currentValue, visited, number + currentValue);
                }
            }
        } else if (sign == '>') {
            for (int currentValue = 0; currentValue < lastValue; currentValue++) {
                if (!visited[currentValue]) {
                    dfs(index + 1, currentValue, visited, number + currentValue);
                }
            }
        }
        visited[lastValue] = false;
    }
}
